File: Trabalhos/T2/partido.cpp
#include <iostream>

#include "partido.hpp"

using namespace std;

Partido::Partido(int numero, const string &sigla) : numero(numero), sigla(sigla), qtdVotosLegenda(0), qtdVotosNominais(0)
{
}

const string &Partido::getSigla() const
{
    return this->sigla;
}

int Partido::getNumero() const
{
    return this->numero;
}

int Partido::getQtdVotosNominais() const
{
    return this->qtdVotosNominais;
}

int Partido::getQtdVotosLegenda() const
{
    return this->qtdVotosLegenda;
}

int Partido::getQtdVotos() const
{
    return this->qtdVotosLegenda + this->qtdVotosNominais;
}

Candidato &Partido::criaCandidato(const int &numeroVotavel, const int &codigoMunicipio, const int &codigoCargo, string &nome, const int &numeroFederacao, const Data &dataNascimento, const int &situacaoCandidatura, const int &genero, const int &foiEleito)
{
    Candidato *c = new Candidato(numeroVotavel, codigoMunicipio, codigoCargo, nome, *this, numeroFederacao, dataNascimento, situacaoCandidatura, genero, foiEleito >= 0 && foiEleito <= 3);

    this->candidatos.insert({numeroVotavel, c});

    return *c;
}

void Partido::adicionaVotosLegendaPartido(const int &qtdVotos)
{
    this->qtdVotosLegenda += qtdVotos;
}

void Partido::adicionaVotosNominaisPartido(const int &qtdVotos)
{
    this->qtdVotosNominais += qtdVotos;
}

const map<int, Candidato *> &Partido::getCandidatos() const
{
    return this->candidatos;
}

const Candidato &Partido::getCandidatoMaisVotado() const
{
    vector<Candidato *> candidatosComVotos;

    for (const auto &pair : this->candidatos)
    {
        candidatosComVotos.push_back(pair.second);
    }

    // Ordena pelos votos dos candidatos (usa operator< de Candidato)
    sort(candidatosComVotos.begin(), candidatosComVotos.end(), CandidatoComparator{});
    // operator< ordena do maior para o menor, então front() é o mais votado
    return *candidatosComVotos.front();
}

const set<Candidato *> Partido::getCandidatosEleitos() const
{
    set<Candidato *> candidatosEleitos;

    for (const auto &pair : this->candidatos)
    {
        if (pair.second->getFoiEleito())
        {
            candidatosEleitos.insert(pair.second);
        }
    }

    return candidatosEleitos;
}

int Partido::getQtdCandidatosEleitos()
{
    return this->getCandidatosEleitos().size();
}

Partido::~Partido()
{
    for (const auto &pair : this->candidatos)
    {
        delete pair.second;
    }
}

void Partido::toString() const
{
    cout << this->sigla << "; "
         << this->numero << "; "
         << this->qtdVotosNominais << "; "
         << this->qtdVotosLegenda << "; "
         << this->getQtdVotos() << endl;

    for (const auto &pair : this->candidatos)
    {
        if (pair.second)
            pair.second->toString();
    }
}

File: Trabalhos/T1/src/Partido.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

public class Partido implements Comparable<Partido> {
    private final int numero; // número do partido do candidato -> NR_PARTIDO;
    private final String sigla; // sigla do partido do candidato -> SG_PARTIDO;
    private int qtdVotosLegenda; // quantidade de votos de legenda do partido;
    private int qtdVotosNominais; // quantidade de votos de todos os candidatos do partido somados;
    private final Map<Integer, Candidato> candidatos = new LinkedHashMap<>(); // lista de candidatos do partido;

    public Partido(int numero, String sigla) {
        this.numero = numero;
        this.sigla = sigla;
        this.qtdVotosLegenda = 0;
    }

    // retorna a sigla do partido;
    public String getSigla() {
        return sigla;
    }

    // retorna o número do partido;
    public int getNumero() {
        return numero;
    }

    // retorna a quantidade de votos de legenda do partido;
    public int getQtdVotosLegenda() {
        return qtdVotosLegenda;
    }

    // retorna a quantidade de votos nominais do partido;
    public int getQtdVotosNominais() {
        return qtdVotosNominais;
    }

    // retorna a quantidade total de votos do partido;
    public int getQtdVotos() {
        return qtdVotosLegenda + qtdVotosNominais;
    }

    // cria e adiciona um candidato ao partido;
    public Candidato criaCandidato(int numeroVotavel, int codigoMuncipio, int codigoCargo,
            String nome, int numeroFederacao, String dataNascimento,
            int situacaoCandidatura, int genero, int foiEleito) {
        Candidato c = new Candidato(numeroVotavel, codigoMuncipio, codigoCargo, nome, this, numeroFederacao,
                dataNascimento, situacaoCandidatura, genero, foiEleito);
        candidatos.put(c.getNumeroVotavel(), c);
        return c;
    }

    // adiciona votos nominais ao partido;
    public void adicionaVotosNominaisPartido(int qtdVotos) {
        this.qtdVotosNominais += qtdVotos;
    }

    // adiciona votos de legenda ao partido;
    public void adicionaVotosLegendaPartido(int qtdVotos) {
        this.qtdVotosLegenda += qtdVotos;
    }

    // retorna todos os candidatos do partido;
    public Map<Integer, Candidato> getCandidatos() {
        return new LinkedHashMap<>(candidatos);
    }

    // retorna o candidato mais votado do partido;
    public Candidato getCandidatoMaisVotado() {
        ArrayList<Candidato> candidatosComVotos = new ArrayList<>();
        candidatosComVotos.addAll(this.candidatos.values());
        Collections.sort(candidatosComVotos);
        return candidatosComVotos.getFirst();
    }

    // retorna o conjunto de candidatos eleitos do partido;
    public Set<Candidato> getCandidatosEleitos() {
        Set<Candidato> candidatosEleitos = new HashSet<>();
        for (Candidato c : this.getCandidatos().values()) {
            if (c.getFoiEleito())
                candidatosEleitos.add(c);
        }
        return candidatosEleitos;
    }

    // retorna a quantidade de candidatos eleitos do partido;
    public int getQtdCandidatosEleitos() {
        return this.getCandidatosEleitos().size();
    }

    // compara partidos por votos totais e, em caso de empate, pelo número;
    @Override
    public int compareTo(Partido p) {
        int diffVotos = p.getQtdVotos() - this.getQtdVotos();
        if (diffVotos == 0) {
            return this.getNumero() - p.getNumero();
        }
        return diffVotos;
    }
}
